File: bot/app/qna.py
#  Copyright 2024 AI RnD Lab
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import os

import aiohttp
from pydantic import BaseModel

# URL сервиса вопросов и ответов
QNA_URL = os.getenv('QNA_URL', 'http://qna:8080')


# Модель ответа
class Answer(BaseModel):
    id: str
    answer: str


# Получение ответа от пайплайна
async def get_answer(question: str, chat_id: str) -> Answer:
    async with aiohttp.ClientSession() as session:
        request = {'question': question, 'chat_id': chat_id}
        async with session.post(f'{QNA_URL}/api/answers', json=request) as response:
            if response.status == 200:
                json = await response.json()
                return Answer(**json)  # Возвращаем объект Answer
            else:
                raise Exception(
                    f"Ошибка получения ответа: {response.status} {await response.text()}")


# Лайкнуть ответ
async def like_answer(answer_id: str) -> None:
    async with aiohttp.ClientSession() as session:
        async with session.post(f'{QNA_URL}/api/answers/{answer_id}/liking') as response:
            if response.status != 200:
                raise Exception(
                    f"Ошибка положительной оценки ответа '{answer_id}': {response.status} {await response.text()}")


# Дизлайкнуть ответ
async def dislike_answer(answer_id: str) -> None:
    async with aiohttp.ClientSession() as session:
        async with session.post(f'{QNA_URL}/api/answers/{answer_id}/disliking') as response:
            if response.status != 200:
                raise Exception(
                    f"Ошибка отрицательной оценки ответа '{answer_id}': {response.status} {await response.text()}")

File: bot/qna.js
const axios = require('axios');
const { qnaUrl } = require('./config');

// URL сервиса вопросов и ответов
const QNA_URL = qnaUrl || 'http://qna:8080';

// Модель ответа
class Answer {
  constructor(id, answer) {
    this.id = id;
    this.answer = answer;
  }
}

// Получение ответа от пайплайна
async function getAnswer(question, chatId) {
  try {
    const request = { question, chatId };
    const response = await axios.post(`${QNA_URL}/api/answers`, request);

    if (response.status === 200) {
      const { id, answer } = response.data;
      return new Answer(id, answer);
    } else {
      throw new Error(`Ошибка получения ответа: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка получения ответа: ${error.message}`);
  }
}

// Лайкнуть ответ
async function likeAnswer(answerId) {
  try {
    const response = await axios.post(`${QNA_URL}/api/answers/${answerId}/liking`);

    if (response.status !== 200) {
      throw new Error(`Ошибка положительной оценки ответа '${answerId}': ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка положительной оценки ответа '${answerId}': ${error.message}`);
  }
}

// Дизлайкнуть ответ
async function dislikeAnswer(answerId) {
  try {
    const response = await axios.post(`${QNA_URL}/api/answers/${answerId}/disliking`);

    if (response.status !== 200) {
      throw new Error(`Ошибка отрицательной оценки ответа '${answerId}': ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка отрицательной оценки ответа '${answerId}': ${error.message}`);
  }
}

async function getProfiles() {
  try {
    const response = await axios.get(`${QNA_URL}/api/profiles`);

    if (response.status === 200) {
      return response.data;
    } else {
      throw new Error(`Ошибка получения профилей: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка получения профилей: ${error.message}`);
  }
}

async function postChats(id, username) {
  try {
    const request = { id, username, type: 'telegram' };
    const response = await axios.post(`${QNA_URL}/api/chats`, request);

    if (response.status !== 200) {
      throw new Error(`Ошибка отправки чата: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка отправки чата: ${error.message}`);
  }
}

async function getInfo(chatId) {
  try {
    const response = await axios.get(`${QNA_URL}/api/chats/${chatId}`);

    if (response.status === 200) {
      return response.data;
    } else {
      throw new Error(`Ошибка получения информации: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка получения информации: ${error.message}`);
  }
}

async function getFaq() {
  try {
    const response = await axios.get(`${QNA_URL}/api/faq`);

    if (response.status === 200) {
      return response.data;
    } else {
      throw new Error(`Ошибка получения часто задаваемых вопросов: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка получения часто задаваемых вопросов: ${error.message}`);
  }
}

async function updateProfile(chatId, profileId) {
  try {
    const response = await axios.patch(`${QNA_URL}/api/chats/${chatId}/profiles/${profileId}`);

    if (response.status === 200) {
      return response.data;
    } else {
      throw new Error(`Ошибка обновления профиля: ${response.status} ${response.statusText}`);
    }
  } catch (error) {
    throw new Error(`Ошибка обновления профиля: ${error.message}`);
  }
}

// Экспортируем функции
module.exports = {
  getAnswer,
  likeAnswer,
  dislikeAnswer,
  getProfiles,
  postChats,
  getInfo,
  getFaq,
  updateProfile,
};
